src/transform/src/analysis/equivalences.rs

```diff
@@ -913,3 +935,41 @@ impl<T: Clone + Ord> UnionFind<T> for BTreeMap<T, T> {
         }
     }
 }
+
+use mz_expr::AggregateFunc;
+/// True iff the aggregate function returns an input datum.
+fn aggregate_is_input(aggregate: &AggregateFunc) -> bool {
```

Comment:
You could consider moving this where all the other property methods (`default`, `identity_datum`, `propagates_nonnull_constraint`, ...) of aggregate functions are defined, i.e., in this block: https://github.com/MaterializeInc/materialize/blob/5f320c115b06d7b93d16a3a55f7b7b9e7a7b45f8/src/expr/src/relation/func.rs#L3078

Reply:
Another thing is that all the other property methods avoid wildcard pattern matches. This is to prevent forgetting about a property function when adding a new aggregate. (This happened at least once in the past.) So, you could consider eliminating the wildcard match also here. (Although, this is not so critical for this particular method, because forgetting about a new aggregate would just lead to missing an optimization, but not to wrong results or panics.)

Reply:
Ah, so .. this is indeed a philosophical question. This is logic that at the moment no one needs / wants other than this transform, and handing it off to the more general implementation section might be an anti-pattern (the "contract" is less clear from the context). I understand wanting to co-locate the things, but also that co-location can be a bit of a dumping ground for "thing I needed, once".

Reply:
Ok, I can totally accept that reasoning.

And what do you think about the wildcard question? Having the method hidden in this file makes it even more likely to be forgotten about when adding a new aggregate.

Reply:
Umm, btw. my IDE is telling me that `ColumnKnowledge` has this exact list of aggregates in a check of its own :)
https://github.com/MaterializeInc/materialize/blob/9a19cc793468c11ba269f44d942c60b688caaa17/src/transform/src/column_knowledge.rs#L399

Although, `ColumnKnowledge` is on its way out, so maybe this doesn't mean much.

Reply:
I'm going to merge as is, but we can pick up the conversation about software engineering principles, and if it turns out we prefer moving the code around should be easy to do.